src/lib/HST/Environment.hs

```diff
@@ -1,101 +1,70 @@
 
--- | An entry of the 'Environment' for a data type whose constructors are in
---   scope.
-data DataEntry a = DataEntry
-  { dataEntryName :: TypeName a
-    -- ^ The name of the data type.
-  , dataEntryCons :: [ConName a]
-    -- ^ The names of the data type's constructors.
-  }
+import           HST.Effect.InputModule
+  ( ConEntry, ConName, ModuleInterface(..), TypeName )
+import qualified HST.Frontend.Syntax    as S
 
 -------------------------------------------------------------------------------
 -- Environment                                                               --
 -------------------------------------------------------------------------------
--- | A data type for the state of the pattern matching compiler.
+-- | A data type for the environment of the pattern matching compiler
+--   containing data types and data constructors currently in scope.
 data Environment a = Environment
-  { envConEntries  :: Map (ConName a) (ConEntry a)
-    -- ^ Maps names of constructors to their 'ConEntry's.
-  , envDataEntries :: Map (TypeName a) (DataEntry a)
-    -- ^ Maps names of data types to their 'DataEntry's.
+  { envCurrentModule   :: ModuleInterface a
+    -- ^ The module interface for the current module, containing its data types
+    --   and data constructors.
+  , envImportedModules :: [(S.ImportDecl a, ModuleInterface a)]
+    -- ^ A list of all successfully imported module interfaces including their
+    --   import declarations. 
+  , envOtherEntries    :: ModuleInterface a
+    -- ^ A module interface containing all other data types and data
+    --   constructors currently in scope.
   }
 
--- | An empty 'Environment'.
-emptyEnv :: Environment a
-emptyEnv
-  = Environment { envConEntries = Map.empty, envDataEntries = Map.empty }
-
 -------------------------------------------------------------------------------
 -- Lookup                                                                    --
 -------------------------------------------------------------------------------
--- | Looks up the entry of a data constructor with the given name in the
---   environment.
+-- | Looks up the constructors belonging to the the given data type name in the
+--   given environment. The result includes the names of the modules the
+--   constructors came from, if available.
 --
---   Returns @Nothing@ if the data constructor is not in scope.
-lookupConEntry :: ConName a -> Environment a -> Maybe (ConEntry a)
-lookupConEntry name = Map.lookup name . envConEntries
+--   Returns an empty list if the data type name is not in scope.
+--   Returns multiple module names and constructor lists if the data type name
+--   is ambiguous.
+lookupConEntries
+  :: TypeName a -> Environment a -> [(Maybe (S.ModuleName a), [ConEntry a])]
+lookupConEntries typeName env = mapMaybe
+  (\x ->
+   fmap (interfaceModName x, ) (Map.lookup typeName (interfaceDataCons x)))
+  (envCurrentModule env
+   : envOtherEntries env
+   : map snd (envImportedModules env))
```

Comment:
The lookup function is quite complicated and is likely to become more complicated when #12 is fixed. In the Free Compiler we've solved this issue by storing all entries (from the current and from imported modules) in a single map in the environment. When a module is imported, the map from its module interface is just merged with the map in the environment. To allow ambiguous imports, sets are stored internally.

```haskell
envConEntries :: Map (ConName a) (Set (ConEntry a))
envDataEntries :: Map (TypeName a) (Set (DataEntry a))
```

Supporting qualified identifiers can then be archived by adding an entry to the map twice (once with the qualified and once with the unqualified name). As a result the lookup function becomes incredibly lightweight.

If you don't want to adopt the approach described above, I'd suggest to add a common implementation for `lookupConEntries` and `lookupTypeName` to reduce code duplication.

Reply:
I think the lookup functions themselves aren't very complicated even after implementing #12. I also have added a common lookup implementation.
What has become very complicated, however, is correctly qualifying the return values of the lookup. That was like opening Pandora's box to me. You will see what I mean in the PR for #12.